### PR TITLE
Changes to Punishments

### PR DESCRIPTION
**Pull request by bumbadadabum**

### CODE REVIEW

File: punishments.js [hunk 1]
```diff
@@ -667,18 +683,30 @@ Punishments.roomBan = function (room, user, expireTime, id, ...rest) {
 };
 
 Punishments.roomBlacklist = function (room, user, expireTime, id, ...rest) {
-	if (!id) id = user.getLastId();
+	if (!id && user) id = user.getLastId();
 
 	if (!expireTime) expireTime = Date.now() + BLACKLIST_DURATION;
 	let punishment = ['BLACKLIST', id, expireTime].concat(rest);
-	Punishments.roomPunish(room, user, punishment);
+	if (user) {
```
Comment (Zarel) [inline punishments.js]: Would definitely attempt to extract `user` from `id` before going down this road.

Comment (bumbadadabum) [inline punishments.js]: The thing is however that I want it to remain possible to blacklistname a school IP and not blacklist the entire school for a year.

Comment (Zarel) [inline punishments.js]: Then write a dissociate command...

Maybe something like /sharedip that marks a user's IP as shared and dissociates it from punishments.

File: chat-plugins/wifi.js [hunk 2]
```diff
@@ -6,6 +6,8 @@
 
 'use strict';
 
+Punishments.roomPunishmentTypes.set('GIVEAWAY_BAN', 'banned from giveaways');
```
Comment (Zarel) [inline chat-plugins/wifi.js]: Let's not use underscores.

Comment (Zarel) [inline chat-plugins/wifi.js]: I'd prefer just `GIVEAWAYBAN`

File: commands.js [hunk 3]
```diff
@@ -2056,6 +2056,37 @@ exports.commands = {
...
+		let reason = parts[1];
+		let targets = parts[0].split(',').map(s => toId(s));
+
+		for (let i = 0; i < targets.length; i++) {
+			let userid = targets[i];
+
+			Punishments.roomBlacklist(room, null, null, userid, reason);
+
+			let thisUser = Users(userid);
+			if (thisUser && thisUser.confirmed && room.isPrivate !== true) {
```
Comment (Zarel) [inline commands.js]: You can check confirmed status for offline users too...

File: punishments.js [hunk 4]
```diff
@@ -117,9 +117,15 @@ Punishments.roomIps = new NestedPunishmentMap();
 //   'LOCK'
 //   'BAN'
 //   'NAMELOCK'
-// For room punishments, they can instead be one of two:
-//   'ROOMBAN'
-//   'BLACKLIST'
+
+// For room punishments, they can be anything in the roomPunishmentTypes map.
```
Comment (Zarel) [inline punishments.js]: You should still list `ROOMBAN` and `BLACKLIST` here

Comment (Zarel) [inline punishments.js]: (to match lock/ban/namelock)

File: commands.js (hunk 3 above)
Comment (bumbadadabum) [inline commands.js]: How?

File: punishments.js (hunk 4 above)
Comment (bumbadadabum) [inline punishments.js]: Will add that

File: chat-plugins/wifi.js (hunk 2 above)
Comment (bumbadadabum) [inline chat-plugins/wifi.js]: Sure.

**Pull request merged**